packages/contracts/contracts/plugin/PluginInstaller.sol

```diff
@@ -0,0 +1,360 @@
+import "@openzeppelin/contracts/proxy/transparent/TransparentUpgradeableProxy.sol";
+import "@openzeppelin/contracts/proxy/ERC1967/ERC1967Proxy.sol";
+
+import {Permission, PluginManager} from "./PluginManager.sol";
+import {PluginERC1967Proxy} from "../utils/PluginERC1967Proxy.sol";
+import {TransparentProxy} from "../utils/TransparentProxy.sol";
+import {bytecodeAt} from "../utils/Contract.sol";
+
+import {PluginUUPSUpgradeable} from "../core/plugin/PluginUUPSUpgradeable.sol";
+import {PluginClones} from "../core/plugin/PluginClones.sol";
+import {Plugin} from "../core/plugin/Plugin.sol";
+import {PluginTransparentUpgradeable} from "../core/plugin/PluginTransparentUpgradeable.sol";
+import {DaoAuthorizable} from "../core/component/DaoAuthorizable.sol";
+
+import {DAO, IDAO} from "../core/DAO.sol";
+import {PluginRepo} from "./PluginRepo.sol";
+import {AragonPluginRegistry} from "../registry/AragonPluginRegistry.sol";
+
+/// @notice Plugin Installer that has root permissions to install plugin on the dao and apply permissions.
+contract PluginInstaller is DaoAuthorizable {
+    using ERC165Checker for address;
+
+    bytes32 public constant INSTALL_PERMISSION_ID = keccak256("INSTALL_PERMISSION");
+    bytes32 public constant SET_REPO_REGISTRY_PERMISSION_ID =
+        keccak256("SET_REPO_REGISTRY_PERMISSION");
+
+    struct InstallPlugin {
+        PluginManager manager;
+        bytes data;
+    }
+
+    struct UpdatePlugin {
+        PluginManager manager;
+        bytes data;
+        address proxy;
+        uint16[3] oldVersion;
+    }
+
+    struct PluginUpdateDeployInfo {
+        address plugin;
+        PluginRepo pluginManagerRepo; // where plugin manager versions are handled.
+        address oldPluginManager;
+        address newPluginManager;
+    }
+
+    mapping(bytes32 => bytes32) private installPermissionHashes;
+    mapping(bytes32 => bool) private pluginInstalledChecker;
+    mapping(bytes32 => bytes32) private updatePermissionHashes;
+    mapping(bytes32 => bytes32) private helperHashes;
+
+    AragonPluginRegistry public repoRegistry;
+
+    error InstallNotAllowed();
+    error PluginCountTooBig();
+    error UpdateNotAllowed();
+    error AlreadyThisVersion();
+    error UpgradeNotExistOnProxy();
+    error NotSupportsUpgradable();
+
+    error PermissionsWrong();
+    error PluginNotDeployed();
+    error HelpersWrong();
+    error PluginRepoNotExists();
+    error PluginWithTheSameAddressExists();
+
+    /// @notice Thrown after the plugin installation to detect plugin was installed on a dao.
+    /// @param dao The dao address that plugin belongs to.
+    /// @param plugin the plugin address.
+    event PluginInstalled(address dao, address plugin);
+
+    event Deployed(
+        address indexed sender,
+        address indexed dao,
+        address indexed plugin,
+        address pluginManager,
+        address[] helpers,
+        Permission.ItemMultiTarget[] permissions,
+        bytes data
```

Comment:
Do we need to emit the `data`?

Reply:
yes, i think it an important information to show how this installation is prepared.